### ISSUE TITLE
Support for "foreign" endianess

### ISSUE THREAD

**Issue by flanfly** [m4b/goblin#1]
Goblin assumes that the integers inside the headers have the same byte order as the current platform. This obv. fails when reading ELF files meant for architectures with different endianess. Do you have any plans to add (or accept PRs) for byte order aware reading of headers?

**Comment by m4b** [comment]
Hi @flanfly ! Thanks for opening an issue :)

Endianness aware parsing would be awesome. I swear I did this for rdr because it can read arm64 binaries but they might still be little endian!

I don't currently have plans to add this functionality, but would definitely accept a PR implementing the functionality, after a discussion and mutual agreement (don't want to waste your time if it doesn't align with your vision).

But first, some considerations and (perhaps) harsh requirements:

1. Should the big endian header/component be a separate struct?
2. If yes, we'll need to feature flag it, because one of the goals of the crate is to have the lib massively configurable (primarily because dryad must have as little relocations as possible, although this might become a moot point if the linker can dump any unused symbols)
3. If no, how invasive would the implementation be? How difficult?
4. This last one might sink your enthusiasm, but I'd prefer the crate to have zero deps. (Or for the endian parser it uses a single crate with no deps)

For last point however, I think we could find a good compromise by adding the endian parsing functionality as a feature flag, and then whatever crate uses it can opt in to get the foreign parser. Or something like that. Hehe, haven't thought too much about it, but always good to!

I dunno, what do you think, what did you have in mind?

:)

**Comment by flanfly** [comment]
Hey @m4b, I started to implement an ELF parser for a disassembler I'm writing but I was told to cut down on the yak shaving :smile:

[The project](https://github.com/das-labor/panopticon) and the [ELF parser](https://github.com/das-labor/panopticon/tree/master/lib/src/elf/parse.rs).

What I did was to define the headers with fields that are large enough for 32 and 64 bit ELF files. The functions that read header structures from disk have additional parameters to specify the byte order and whenever to [read the 32 or 64 variants of words, xwords et.al](https://github.com/das-labor/panopticon/blob/master/lib/src/elf/parse.rs#L27)

I only took a brief look but changing the `from_fd` function to [read all fields manually](https://github.com/das-labor/panopticon/blob/master/lib/src/elf/parse.rs#L748) using methods from the `byteorder` crate like I did in [here] and changing the `from_path` function to call the `from_fd` functions with the right parameters should suffice.

If you're hell-bent on the zero deps issue I can write the needed `ntoh` functions myself to avoid the dependency on `byteorder`.

**Comment by m4b** [comment]
I'm aware of the project (keep up the good work, excited to see where it goes! 😃)

So I see what you've done in your implementation.

Unfortunately that particular approach won't work, as all of my structs have `repr(C)`, as they're used in FFI, binary loading, and gdb accesses, etc., and must have that exact layout in memory, as I have to use them for pointer types, casts, memory transmutes and lots of other unsafe activities which require the specific 32/64 bit layout(s).

I was going to add elf32 struct variants at some point later, and leave it up to clients of the lib or a downstream wrapper to deal with "automatically" dealing with one representation of a binary using an adapter or something else, similar to what you've done.

This single representation was also the original use case of the "goblin" format, which isn't defined here (yet).

Anyway, that's the first problem, the structs cannot change, and 32 bit variants (when I write them, not on top of bucket list right now, because not urgent) will be separate c layout structs (and feature switchable).

I looked at byteorder and it seems fine; it has no deps, and code quality looks good.  I'd prefer to enable it with a feature flag, if possible. I will experiment adding it to goblin, and seeing how many relocations it adds to dryad (my guess is it might be OK?)

So that's perhaps the second question/problem: can we add endianness parsing without changing the structs and the API ? (Of course minor changes are expected, but I'd prefer as little as possible downstream changes, ideally only to the fd readers as they should only be the APIs affected by endianness concerns since if you're reading in memory, what's in memory had better match your host machine ;) )

So:
1. I think byteorder is fine to use
2. Most of current ELF API can't change, modulo fd forms
3. Can we feature flag this?

What do you think?

**Comment by m4b** [comment]
@flanfly 

So I've pushed a sample https://github.com/m4b/goblin/tree/endian branch which might start off something like I mean with feature flags:

https://github.com/m4b/goblin/commit/56f793b05746f60dcaa2541e7a5dfaeafca21ffc

If you decide you feel like making a PR, might be a good place to start?

I've also realized I want goblin's features to be all "opt-in" by default, which means the features need to be "opt-out", as Cargo only supports opting into features individually.  This means prefixing "no_", etc.

For example, here is dryad's new `Cargo.toml` for the sample `no_endian_fd` feature gate:

```
[dependencies.goblin]
path = "../goblin"
default-features = false
features = ["no_mach", "no_mach32", "no_pe", "no_pe32", "no_goblin", "no_endian_fd"]
```

This suffices to:

1. not download byteorder when compiling
2. not use the endian implementation of the fds
3. opt out of everything except elf

Since dryad's use-case is unusual in only wanting a specific feature, and most people will want to use the different binary targets in the crate, I decided the `no` prefix was the best way to go.

So unfortunately, for now the endianness does need to be like this, because I'm getting linking errors in dryad when it attempts to use the endian implementation:

```
target/x86_64-unknown-linux-musl/debug/deps/libgoblin-a5e455b60d4df90b.rlib(goblin-a5e455b60d4df90b.0.o): In function `byteorder::{{impl}}::read_u16':
/home/m4b/.cargo/registry/src/github.com-1ecc6299db9ec823/byteorder-0.5.3/src/lib.rs:2: undefined reference to `_$LT$BigEndian$u20$as$u20$ByteOrder$GT$::read_u16::_FILE_LINE::h0a2b2ba52f0892c7'
/home/m4b/.cargo/registry/src/github.com-1ecc6299db9ec823/byteorder-0.5.3/src/lib.rs:2: undefined reference to `_$LT$BigEndian$u20$as$u20$ByteOrder$GT$::read_u16::_FILE_LINE::h96e2ea3736c2e3c0'
```

(which also means it will increase the number of relocations, but likely not by many)

I don't know why this would happen, since I'm able to use other crates, but this is the first time i'm using a crate which itself has dependencies, so likely something weird going on during linking.

In any event I don't feel like thinking about it right now, so if we go forward, the endian implementation will need to be feature gated.  It should be as simple as leaving the old implementation with a feature flag like in the header sample, and then adding the byteorder manual reader, and filling up the structs that way, I should think?
